test/test_dataloader.py

```diff
@@ -1846,6 +1871,85 @@ def test_pin_memory(self):
             self.assertTrue(sample['another_dict']['a_number'].is_pinned())
 
 
+class DummyDataset(torch.utils.data.Dataset):
+    def __init__(self):
+        self.data = list(range(10))
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        if torch.is_tensor(idx):
+            idx = idx.tolist()
+        # The persistent workers always maintain the original
+        # dataset through the dataloader lifetime
+        # so the attributes will remain the same as the
+        # first time the workers where spawned (dataloader iteration)
+        assert self.start == 0
+        return self.data[idx]
+
+
+@unittest.skipIf(
+    TEST_WITH_TSAN,
+    "Fails with TSAN with the following error: starting new threads after multi-threaded "
+    "fork is not supported. Dying (set die_after_fork=0 to override)")
+class TestDataLoaderPersistentWorkers(TestDataLoader):
+
+    def setUp(self):
+        super(TestDataLoaderPersistentWorkers, self).setUp()
+        self.persistent_workers = True
+
+    @unittest.skipIf(IS_SANDCASTLE, "subprocess doesn't work in FB internal CI")
+    @unittest.skipIf(IS_WINDOWS, "No 'resource' module on Windows")
+    def test_fd_limit_exceeded(self):
+        # See NOTE [ DataLoader on Linux and open files limit ]
+        import subprocess
+        subprocess.check_output([sys.executable, '-c', """\
+import torch
+import resource
+from torch.utils.data import DataLoader, IterableDataset
+
+class RandomDataset(IterableDataset):
+    def __init__(self, len, size):
+        super(RandomDataset).__init__()
+        self.len = len
+        self.size = size
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.len <= 0:
+            raise StopIteration
+        self.len -= 1
+        return torch.randn(self.size)
+
+try:
+    keep_fds_alive = []
+    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))
+    for random_t in DataLoader(RandomDataset(200, (2,2)),
+                               num_workers=1, persistent_workers=True):
+      random_t.max(dim=0)
+      keep_fds_alive.append(random_t)
+except RuntimeError as e:
+    assert "ulimit -n" in str(e)
+    assert "set_sharing_strategy" in str(e)
+"""])
+
+    def test_dataset_not_reset(self):
+        dataset = DummyDataset()
+        dataloader = self._get_data_loader(dataset, num_workers=2)
+        dataset.start = 0
+        for i in range(10):
+            for x in dataloader:
+                pass
+            # Changing the start value here doesn't have any effect in the dataset
+            # cached by the workers. since they are not recreated between epochs
+            # and can cache values safely
+            dataset.start = i
```

Comment:
It is failing nicely with `self.persistent_workers = False` ?

Reply:
It is confirmed to fail (which is expected), but that it is actually checking is the fact that we are not making any new copies of `dataset`. There is no test which checks that you are using exactly same DataSet objects after reset. 

It is still acceptable to merge as is anyway.

Reply:
Actually, the `DataSet` object is instantiated only once, by the user, and then it seems that is cloned when calling:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L764-L769

The object seems to be hardly cloned in the new process memory space, no pickle (I tried to verify this by overriding `__getstate__` or `__setstate__`, additionally, in the new processes, the object id is the same than in the main process, so there is no actual "instantiation".

It seems that the only way we have to assert that the object is not being recreated is actually this test. If the object is recreated it will get new values in the next iteration (.start) and if not, it will retain the cached values.